ethexe/service/src/fast_sync.rs

```diff
@@ -0,0 +1,543 @@
+                }
+                RequestMetadata::Mailbox => {
+                    let mailbox: Mailbox =
+                        Decode::decode(&mut &data[..]).expect("`db-sync` must validate data");
+                    for user_mailbox in mailbox.as_ref().values() {
+                        manager.add(user_mailbox.hash(), RequestMetadata::Data);
+                    }
+                }
+                RequestMetadata::Data => {}
+            }
+        }
+    }
+
+    log::info!("Network data getting is done");
+}
+
+async fn instrument_codes(
+    db: &Database,
+    compute: &mut ComputeService,
+    mut code_ids: HashSet<CodeId>,
+) -> Result<()> {
+    if code_ids.is_empty() {
+        log::info!("No codes to instrument. Skipping...");
+        return Ok(());
+    }
+
+    log::info!("Instrument {} codes", code_ids.len());
+
+    for &code_id in &code_ids {
+        let code_info = db
+            .code_blob_info(code_id)
+            .expect("observer must fulfill database");
+        let original_code = db
+            .original_code(code_id)
+            .expect("observer must fulfill database");
+        compute.receive_code(code_id, code_info.timestamp, original_code);
+    }
+
+    while let Some(event) = compute.next().await {
+        if let ComputeEvent::CodeProcessed(CodeCommitment { id, .. }) = event? {
+            code_ids.remove(&id);
+            if code_ids.is_empty() {
+                break;
+            }
+        }
+    }
+
+    log::info!("Instrumentation done");
+    Ok(())
+}
+
+pub(crate) async fn sync(service: &mut Service) -> Result<()> {
+    let Service {
+        observer,
+        compute,
+        network,
+        db,
+        sender,
+        ..
+    } = service;
+    let Some(network) = network else {
+        log::warn!("Network service is disabled. Skipping fast synchronization...");
+        return Ok(());
+    };
+
+    log::info!("Fast synchronization is in progress...");
+
+    let finalized_block = sync_finalized_head(observer).await?;
+    let Some(EventData {
+        program_states,
+        program_code_ids,
+        needs_instrumentation_codes,
+        latest_committed_block,
+        previous_committed_block,
+    }) = EventData::collect(db, finalized_block).await?
+    else {
+        log::warn!("No any committed block found. Skipping fast synchronization...");
+        return Ok(());
+    };
+
+    instrument_codes(db, compute, needs_instrumentation_codes).await?;
+
+    let latest_block_header = OnChainStorage::block_header(db, latest_committed_block)
+        .expect("observer must fulfill database");
+
+    sync_from_network(network, db, &program_states).await;
+
+    let schedule =
+        ScheduleRestorer::from_storage(db, &program_states, latest_block_header.height)?.restore();
+
+    for (program_id, code_id) in program_code_ids {
+        db.set_program_code_id(program_id, code_id);
+    }
+
+    // NOTE: there is no invariant that fast sync should recover queues
```

Comment:
it would be great if you'd implement or open todo for doing something similar like ScheduleRestorer, but for checking invariants so we could run it in tests to verify that we didn't forget to do something here